osbs/core.py

```diff
@@ -382,6 +383,53 @@ def wait_for_new_build_config_instance(self, build_config_id, prev_version):
 
         raise OsbsResponseException("New BuildConfig instance not found")
 
+    def stream_logs(self, build_id):
+        """
+        stream logs from build
+
+        :param build_id: str
+        :return: iterator
+        """
+        kwargs = {'follow': 1}
+
+        # If connection is closed within this many seconds, give up:
+        min_idle_timeout = 60
+
+        # Stream logs, but be careful of the connection closing
+        # due to idle timeout. In that case, try again until the
+        # call returns more quickly than a reasonable timeout
+        # would be set to.
+        last_activity = time.time()
+        while True:
+            buildlogs_url = self._build_url("builds/%s/log/" % build_id,
+                                            **kwargs)
+            response = self._get(buildlogs_url, stream=1,
+                                 headers={'Connection': 'close'})
+
+            try:
+                check_response(response)
+            except OsbsNetworkException as ex:
+                if ex.status_code == pycurl.FOLLOWLOCATION:
+                    # This is how pycurl reports 'empty reply from server'
+                    # Handle this as though there were no lines returned.
+                    pass
+
+                raise
+            else:
+                for line in response.iter_lines():
+                    last_activity = time.time()
+                    yield line
+
+            idle = time.time() - last_activity
+            logger.debug("connection closed after %ds", idle)
+            if idle < min_idle_timeout:
+                # Finish output
+                raise StopIteration
+
+            since = int(idle - 1)
```

Comment:
That's the intention, yes... but I think we need to subtract 1.

`sinceSeconds` is how many seconds in the past to start streaming logs from.

```
[last_activity] last log message received
:
: <-- resume from anywhere in this period
:
[last_activity + idle] connection closed
```

Since `idle` is the number of seconds since `last_activity`, I think we want to start streaming no more than `idle` seconds in the past (otherwise we'll get duplicated log lines).

We also don't want to miss any lines, but we know that there weren't any for up to `idle` seconds after `last_activity`.

The safest number of seconds in the past to start streaming from seems to be the `idle - 1`, i.e. 1s after the last message we saw.